apps/cowswap-frontend/src/utils/orderUtils/calculateAmountForRate.ts

```diff
@@ -28,15 +28,16 @@ export function calculateAmountForRate({
   const parsedValue = FractionUtils.adjustDecimalsAtoms(
     amount,
     field === Field.INPUT ? inputDecimals : outputDecimals,
-    field === Field.INPUT ? outputDecimals : inputDecimals
+    field === Field.INPUT ? outputDecimals : inputDecimals,
   )
+  const activeRateAsFraction = activeRate instanceof Price ? FractionUtils.fromPrice(activeRate) : activeRate
```

Comment:
The definition of Price is:

```
export declare class Price<TBase extends Currency, TQuote extends Currency> extends Fraction {
```

So a price is a valid fraction, why do we need specific handling? 

Would be the problem that we use in the math the `quotient` instead of multiplying by numerator and dividing by denominator for the INPUT case, and the inverse of that for the OUTPUT case

Reply:
@anxolin thanks for the question!
I've added a test with an explnation how it works:
https://github.com/cowprotocol/cowswap/pull/5322/commits/7f7a81feaefd642030e4135405361bc69607441d

Please let me know if it doesn't answer on your question.

Reply:
Well, it creates more questions to me :) 

I really don't get the test `calculateAmountForRate`.

You multiply 1000000 weis of Ether (1e-12 Ether) by the price of ether --> 3208940687  weis of USDC for 1e18 weis of WETH --> 3208.940687 USDC/ETH

If you multiply you get 3208940687000000 / 1e18 ---> 0.003208940687

Because JSBI is a BigInt library, it will return 0

So maybe the issue, is we "expect" the price to be in units and it's in wei, but this is not how it stores the information inside. However, I believe it does the conversion you seek inside the Price implementation, IF ONLY it was a public method https://github.com/Uniswap/sdk-core/blob/main/src/entities/fractions/price.ts#L77

Reply:
As far as I can tell by their implementation of the Price, its correct. It makes sure you multiply using the right units. So here when you multiply weiWETH value for a price in `weiUSD/weiWETH`, you naturally get weiUSDC.

So I think multiplying directly should be actually safer. You just need to not use later the quotient without adjusting the units

Basically the function I share above (`adjustedForDecimals`) I think should convert a fraction with units of `weiUSD/weiWETH` into `USD/WETH`